factory-containers/assemble-system-image

```diff
@@ -0,0 +1,151 @@
+#!/usr/bin/python3
+#
+# Copyright (c) 2019 Foundries.io
+# SPDX-License-Identifier: Apache-2.0
+#
+import subprocess
+import os
+import requests
+import json
+import argparse
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+def get_targets_from_api(factory: str, build_num: str, token: str):
+    api_base_url = 'https://api.foundries.io/ota/repo/'
+    targets_endpoint = 'api/v1/user_repo/targets.json'
+
+    res_targets = []
+    target_url = os.path.join(api_base_url, factory, targets_endpoint)
+    target_resp = requests.get(target_url, headers={'OSF-TOKEN': token})
+    target_resp.raise_for_status()
+
+    for target_name, target in target_resp.json()['signed']['targets'].items():
+        custom = target.get('custom')
+        if not custom:
+            continue
+
+        if custom.get('version', '') == build_num and custom.get('targetFormat', 'NONE') == 'OSTREE':
+            res_targets.append({target_name: target})
+
+    return res_targets
+
+
+# ${IMAGE_BASENAME}-${MACHINE}.wic.gz
+# https://api.foundries.io/projects/msul-dev01/lmp/builds/224/runs/raspberrypi3-64/lmp-factory-image-raspberrypi3-64.wic.gz
+def get_system_image_from_ci(base_uri: str, image_name: str, image_machine: str, token: str):
+    base_uri = base_uri.replace('ci.foundries.io', 'api.foundries.io')
+    target_image_file = '{}-{}.wic.gz'.format(image_name, image_machine)
+    image_url = os.path.join(base_uri, 'runs', image_machine, target_image_file)
+    logger.info('Fetching ' + image_url)
+    image_resp = requests.get(image_url, headers={'OSF-TOKEN': token})
+    image_resp.raise_for_status()
+    with open(target_image_file, 'wb') as image_file:
+        for data_chunk in image_resp.iter_content(chunk_size=65536):
+            image_file.write(data_chunk)
+
+    subprocess.check_call(['gunzip', '-f', target_image_file])
+    return target_image_file.rstrip('.gz')
+
+
+def get_args():
+    parser = argparse.ArgumentParser('''Add container images to a system image''')
+
+    parser.add_argument('-f', '--factory', help='Factory', default=os.environ.get('FACTORY'))
+    #parser.add_argument('-p', '--project', help='Factory project <factory>/lmp', default=os.environ.get('H_PROJECT'))
+    parser.add_argument('-b', '--build-num', help='Build number', default=os.environ.get('H_BUILD'))
+    parser.add_argument('-t', '--token-file', help='A file containing OSF token', default='/secrets/osftok')
+    parser.add_argument('-w', '--wic-tool', help='A path to WIC utility', default=os.environ.get('WIC_TOOL'))
+    parser.add_argument('-a', '--app-image-dir', help='A path to directory that contains app container images',
+                        default=os.environ.get('APP_IMAGE_DIR'))
+    parser.add_argument('-o', '--out-image-dir', help='A path to directory to put a resultant image to',
+                        default=os.environ.get('OUT_IMAGE_DIR'))
+
+    args = parser.parse_args()
+
+    if args.factory is None:
```

Comment:
That's exactly what I did in the first place, but it doesn't do what I need. If `required=True` then it will yield and an error even if the environment variables are defined.

Reply:
ah - i see why. required=True and a default fight each other. okay.